xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-sources/src/main/java/org/xwiki/notifications/sources/internal/DefaultNotificationParametersFactory.java

```diff
@@ -402,27 +413,32 @@ private void dontUseUserPreferences(NotificationParameters notificationParameter
             .filter(filter -> !excludedFilters.contains(filter.getName())).collect(Collectors.toList());
 
         enableAllEventTypes(notificationParameters);
-        handleLocationParameter(parameters.get(ParametersKey.PAGES),
-            notificationParameters, NotificationFilterProperty.PAGE);
-        handleLocationParameter(parameters.get(ParametersKey.SPACES),
-            notificationParameters, NotificationFilterProperty.SPACE);
 
         String wikis = parameters.get(ParametersKey.WIKIS);
         String currentWiki = parameters.get(ParametersKey.CURRENT_WIKI);
 
         if (StringUtils.isEmpty(currentWiki)) {
             currentWiki = wikiDescriptorManager.getCurrentWikiId();
         }
-        handleLocationParameter(wikis, notificationParameters, NotificationFilterProperty.WIKI);
+
+        handleLocationParameter(parameters.get(ParametersKey.PAGES),
+            notificationParameters, NotificationFilterProperty.PAGE, currentWiki);
+        handleLocationParameter(parameters.get(ParametersKey.SPACES),
+            notificationParameters, NotificationFilterProperty.SPACE, currentWiki);
+        handleLocationParameter(wikis,
+            notificationParameters, NotificationFilterProperty.WIKI, currentWiki);
+
         // When the notifications are displayed in a macro in a subwiki, we assume they should not contain events from
         // other wikis (except if the "wikis" parameter is set).
         // The concept of the subwiki is to restrict a given domain of interest into a given wiki, this is why it does
         // not make sense to show events from other wikis in a "timeline" such as the notifications macro.
         // TODO: add a "handleAllWikis" parameter to disable this behaviour
         // Note that on the main wiki, which is often a "portal" for all the others wikis, we assure it's OK to display
         // events from other wikis.
-        if (StringUtils.isBlank(wikis) && !StringUtils.equals(currentWiki, wikiDescriptorManager.getMainWikiId())) {
-            handleLocationParameter(currentWiki, notificationParameters, NotificationFilterProperty.WIKI);
+        if (StringUtils.isBlank(wikis) && StringUtils.isBlank(parameters.get(ParametersKey.PAGES)) && StringUtils.isBlank(parameters.get(ParametersKey.SPACES))
```

Comment:
This condition clearly miss explanations: I remember we discuss it but I'll need to check back the issue to understand why it's done like that, so it's not clear enough :) 
And AFAICS it doesn't repect the code style: the line is too long. 
I suggest to extract the condition as a private method that you can explain so we all agree on it.

Reply:
I have extracted adding the extra filter into a separate method ``handleSubwikiWithoutLocationParameters`` and added the existing comment as javadoc to that method, extended a bit about why the condition is as it is there.

Unfortunately I got a conflict in the notifications sources ``pom.xml`` when trying to rebase to the latest master commits, and it seems I resolved this the wrong way so github thinks it cannot be merged automatically because of this.

Reply:
Looks ok on Github.